### PR TITLE
Fix links to nprofiles and nevents in notes, and deep linking of all kinds

### PR DESCRIPTION
**Pull request by joshuatbrown**
## Issues covered
#1234
#1231 
#1295 

## Description
This fixes issues with nprofiles and nevents:
- Fixes the issue with some nprofile references by removing the upper limit in the regex (#1234)
- Fixes the issue with some nevent references by removing the upper limit in the regex (#1231)
- Fixes deep linking to a nprofile or nevent (or npub or note) (#1295)

It also replaces older Bech32 decoding with NostrIdentifier decoding.

### Linking to an nprofile in a note
1. Navigate to Discover
2. Tap search
3. Paste note1ar8gnwj6ugxyq6w9aqur74s2sn4dttk6zx0d25p34fsg476kqrksmkqqnr
4. Observe note

Expected: the note should have a link at the beginning which will take the user to the profile for brugeman.

You can also test the following notes that should link to a profile:

- note1alva8prc6ae0r973erz25fg5pcdfs8pq5gnfhzefvylf6zzv2vvsys344q
- note1w6u6stxxvr78lat6d9sknsac0jnpj7ylpm6sfjxgmg7euulpczyq5asrw2

### Linking to an nevent in a note

1. Navigate to Discover
2. Tap search
3. Paste note1qq7am022xm40yrj9k4agcfqwre4lflsq6mytky0suxunrk7ldf5sjwth8x
4. Observe note

Expected: the note from fiatjaf should have a _link_ to a note by 0xtr. Previously it just displayed plain text of the nevent string.

You can also test the following notes that should show the "🔗 Link to note":

- note1ypppzcue3svj2p0l80vp4lf52j3xmykn8yzjdv3gnq2sm4ljp5qqrqp9hd
- note16fddcyxfldwre2zywr96fnkmk7n3rtf4ehntdwy8ltyqgfntwfnq0dm347

### Deep linking to notes and profiles
See #1295 for details on testing deep links. But I suppose I'll reproduce the good parts here. These should all link to something in the Dev version of the app; for Staging, replace `nos-dev:` with `nos-staging:` (or just go to the issue and get the staging links there):

- nos-dev://npub1qfn650vj62j8ntttenwxlem9wqmaa26tw75thn7nvcasamceddvqy2zdu4
- nos-dev://note1w6u6stxxvr78lat6d9sknsac0jnpj7ylpm6sfjxgmg7euulpczyq5asrw2
- nos-dev://nprofile1qqsrhuxx8l9ex335q7he0f09aej04zpazpl0ne2cgukyawd24mayt8gpp4mhxue69uhhytnc9e3k7mgpz4mhxue69uhkg6nzv9ejuumpv34kytnrdaksjlyr9p

I'm not sure what's up with nevents, but they don't work for me. I'm not concerned since this isn't a high-priority issue.

### Linking to an nprofile
| Before | After |
|--------|--------|
| Screenshot 2024-07-10 at 1 48 29 PM | Screenshot 2024-07-10 at 1 48 01 PM | 

### Linking to an nevent
| Before | After |
|--------|--------|
| Simulator Screenshot - iPhone 15 Pro - 2024-07-11 at 13 17 22 | Simulator Screenshot - iPhone 15 Pro - 2024-07-11 at 13 16 29 |

### Deep linking

https://github.com/planetary-social/nos/assets/59564/0a731932-6d2a-459b-897e-d351f0f26f28

### CODE REVIEW

File: Nos/Models/NoteParser.swift [hunk 1]
```diff
@@ -114,7 +115,7 @@ struct NoteParser {
     /// Replaces Nostr entities embedded in the note (without a proper tag) with markdown links
     private func replaceNostrEntities(in content: String) -> String {
         // swiftlint:disable opening_brace operator_usage_whitespace closure_spacing comma superfluous_disable_command
-        let unformattedRegex = /(?:^|\s)@?(?:nostr:)?(?<entity>((npub1|note1|nprofile1|nevent1)[a-zA-Z0-9]{58,255}))/
+        let unformattedRegex = /(?:^|\s)@?(?:nostr:)?(?<entity>((npub1|note1|nprofile1|nevent1)[a-zA-Z0-9]{58,}))/ // TODO: where'd we get the 58? is that right?
```
Comment (joshuatbrown) [inline Nos/Models/NoteParser.swift:118]: @martindsq @mplorentz Do either of you know why we have 58 here? I haven't found anything in the NIP-19 documentation that says, but maybe this is from some other source.

Comment (martindsq) [reply Nos/Models/NoteParser.swift:117]: It is the length of an `npub10elfcs4fr0l0r8af98jlmgdh9c8tcxjvz9qkw038js35mp4dma8qzvjptg` after you get rid of `npub1` which is already captured.

Comment (joshuatbrown) [reply Nos/Models/NoteParser.swift:117]: Is that the shortest possible length? This also looks for note, nprofile, and nevent, so it's not just npub, and I'd like to be sure we get this right. The upper limit of 255 was wrong and causing the bug that's fixed in this PR. Is it possible that npub or any of the others could be shorter?

Comment (mplorentz) [reply Nos/Models/NoteParser.swift]: That is indeed the shortest possible length 👍

**Review by mplorentz — APPROVED** [review]
This works perfectly for me! 

This wasn't asked for in any of the tickets, but as a user I expect to be able to paste all the NIP-19 entities into the search box on the discover tab and have them open the correct thing. I'll bring this up in Slack.

File: Nos/Models/NoteParser.swift (hunk 1 above)
Comment (joshuatbrown) [reply Nos/Models/NoteParser.swift]: Awesome -- thanks!

**Pull request merged**